smack-core/src/main/java/org/jivesoftware/smack/altConnections/HttpLookupMethod.java

```diff
@@ -0,0 +1,218 @@
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.jivesoftware.smack.util.MultiMap;
+import org.jivesoftware.smack.util.PacketParserUtils;
+import org.jivesoftware.smack.xml.XmlPullParser;
+import org.jivesoftware.smack.xml.XmlPullParserException;
+
+import org.jxmpp.jid.DomainBareJid;
+
+/**
+ * <p><i>The HTTP lookup method uses web host metadata to list the URIs of alternative connection methods.</i></p>
+ *
+ * <p>In order to obtain host-meta XRD element from the host in the form of an <i>InputStream</i>,
+ * use {@link #getXrdStream(DomainBareJid)} method. To obtain endpoints for Bosh or Websocket
+ * connection endpoints from host, use {@link LinkRelation#BOSH} and {@link LinkRelation#WEBSOCKET}
+ * respectively with the {@link #lookup(DomainBareJid, LinkRelation)} method. In case one is looking
+ * for endpoints described by other than BOSH or Websocket LinkRelation, use the more flexible
+ * {@link #lookup(DomainBareJid, String)} method.</p>
+ * Example:<br>
+ * <pre>
+ * {@code
+ * List<URI> endpoints = HttpLookupMethod.lookup(DomainBareJid, LinkRelation.WEBSOCKET);
+ * }
+ * </pre>
+ * @see <a href="https://xmpp.org/extensions/xep-0156.html#http">
+ *     HTTP Lookup Method from XEP-0156.
+ *     </a>
+ */
+public final class HttpLookupMethod {
+
+    private static final int BUFFER_SIZE = 256;
+    private static final String XRD_NAMESPACE = "http://docs.oasis-open.org/ns/xri/xrd-1.0";
+
+    private static int contentLength;
+    private static HttpURLConnection httpURLConnection;
+    private static MultiMap<DomainBareJid, MultiMap<String, URI>> CACHE = new MultiMap<>();
+
+    /**
+     * Specifies a link relation for the selected type of connection.
+     */
+    enum LinkRelation {
+        /**
+         * Selects link relation attribute as "urn:xmpp:alt-connections:xbosh".
+         */
+        BOSH("urn:xmpp:alt-connections:xbosh"),
+        /**
+         * Selects link relation attribute as "urn:xmpp:alt-connections:websocket".
+         */
+        WEBSOCKET("urn:xmpp:alt-connections:websocket");
+        private final String attribute;
+        LinkRelation(String relAttribute) {
+            this.attribute = relAttribute;
+        }
+    }
+
+    /**
+     * Constructs a HTTP connection with the host specified by the DomainBareJid
+     * and retrieves XRD element in the form of an InputStream.
+     *
+     * @param xmppServiceAddress address of host
+     * @return InputStream containing XRD element
+     * @throws IOException exception due to input/output operations
+     */
+    public static InputStream getXrdStream(DomainBareJid xmppServiceAddress) throws IOException {
+        final String metadataUrl = "https://" + xmppServiceAddress.toString() + "/.well-known/host-meta";
+        final URL putUrl = new URL(metadataUrl);
+        httpURLConnection = (HttpURLConnection) putUrl.openConnection();
```

Comment:
Do you need to cast it to `HttpURLConnection`? Can't it just stay an `URLConnection`?

Reply:
Changed `HttpURLConnection` to its superclass `URLConnection` :)